File: flutter/android/src/main/kotlin/com/ary/push/flutter/ARYPushPlugin.kt
package com.ary.push.flutter

import android.content.Context
import com.ary.push.ARYPush
import com.ary.push.model.PushNotification
import com.ary.push.model.PushPermissionStatus
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * The Android half of the Flutter bridge.
 *
 * Deliberately thin. Every decision about FCM, rendering, tokens, storage and deduplication is
 * made by the native SDK; this class translates method calls and forwards events. There is no
 * second notification engine here, because two engines would drift and the Dart one could not
 * run when the app is terminated.
 *
 * Engine lifecycle is the subtle part. A hot restart, a hot reload or a recreated
 * `FlutterEngine` tears down the Dart side while the native SDK keeps running, so listeners are
 * attached exactly once per engine attachment and removed on detach. Without that, every hot
 * restart would leave another listener behind and the app would show duplicate notifications
 * that disappear on a cold start.
 */
public class ARYPushPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: Context

    private var eventSink: EventChannel.EventSink? = null

    /**
     * Events that occurred before Dart was listening.
     *
     * Bounded, because a device that receives many messages before the engine attaches must not
     * accumulate them forever. Notification opens are not buffered here: the native SDK persists
     * those itself, which is what makes them survive a terminated launch rather than merely a
     * slow one.
     */
    private val pendingEvents = ArrayDeque<Map<String, Any?>>()

    private val receivedListener: (PushNotification) -> Unit = { notification ->
        emit("received", notification.toMap())
    }

    private val openedListener: (PushNotification) -> Unit = { notification ->
        emit("opened", notification.toMap())
    }

    private val tokenListener: (String) -> Unit = { token ->
        emit("token", mapOf("token" to token))
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // Detaching without removing these would leave the previous engine's listeners alive
        // and produce duplicate events after every hot restart.
        detachListeners()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        pendingEvents.clear()
    }

    // ------------------------------------------------------------------ events

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        attachListeners()

        // Anything that happened while Dart was not listening is delivered now, in order.
        while (pendingEvents.isNotEmpty()) {
            events?.success(pendingEvents.removeFirst())
        }
    }

    override fun onCancel(arguments: Any?) {
        detachListeners()
        eventSink = null
    }

    private fun attachListeners() {
        // The SDK's listener collections ignore duplicates, so re-attaching after a hot restart
        // cannot register the same listener twice.
        ARYPush.addNotificationReceivedListener(receivedListener)
        ARYPush.addNotificationOpenedListener(openedListener)
        ARYPush.addTokenRefreshListener(tokenListener)
    }

    private fun detachListeners() {
        ARYPush.removeNotificationReceivedListener(receivedListener)
        ARYPush.removeNotificationOpenedListener(openedListener)
        ARYPush.removeTokenRefreshListener(tokenListener)
    }

    private fun emit(type: String, payload: Map<String, Any?>) {
        val event = mapOf("type" to type, "payload" to payload)
        val sink = eventSink
        if (sink == null) {
            if (pendingEvents.size >= MAX_PENDING_EVENTS) pendingEvents.removeFirst()
            pendingEvents.addLast(event)
            return
        }
        sink.success(event)
    }


    // ------------------------------------------------------------------ method calls

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                ARYPush.initialize(
                    applicationContext,
                    FlutterConfigMapper.from(call.arguments)
                )
                result.success(null)
            }

            "isInitialized" -> result.success(ARYPush.isInitialized)

            // Answered asynchronously: the system prompt is shown and the result reported once
            // the user has decided.
            "requestPermission" ->
                ARYPush.requestPermission { status -> result.success(status.toWire()) }

            "getPermissionStatus" ->
                result.success(ARYPush.getPermissionStatus().toWire())

            "openNotificationSettings" -> {
                ARYPush.openNotificationSettings(applicationContext)
                result.success(null)
            }

            "getInstallationId" -> result.success(ARYPush.getInstallationId())

            "getPushToken" -> ARYPush.getPushToken { token -> result.success(token) }

            "getPushProvider" -> result.success("fcm")

            "login" -> {
                val userId = call.argument<String>("userId")
                if (userId.isNullOrBlank()) {
                    result.error("invalid_argument", "userId must not be blank", null)
                } else {
                    ARYPush.login(userId)
                    result.success(null)
                }
            }

            "logout" -> {
                ARYPush.logout()
                result.success(null)
            }

            "getUserId" -> result.success(ARYPush.getUserId())

            "addTags" -> {
                ARYPush.addTags(call.stringMap("tags"))
                result.success(null)
            }

            "removeTags" -> {
                ARYPush.removeTags(call.stringList("keys").toSet())
                result.success(null)
            }

            "removeAllTags" -> {
                ARYPush.removeAllTags()
                result.success(null)
            }

            "getTags" -> result.success(ARYPush.getTags())

            "subscribeToTopic" ->
                ARYPush.subscribeToTopic(call.argument<String>("topic").orEmpty()) { ok ->
                    result.success(ok)
                }

            "unsubscribeFromTopic" ->
                ARYPush.unsubscribeFromTopic(call.argument<String>("topic").orEmpty()) { ok ->
                    result.success(ok)
                }

            "getSegments" ->
                // Answered asynchronously: membership is read from the backend.
                ARYPush.getSegments { segments -> result.success(segments.map { it.toMap() }) }

            "getSubscribedTopics" ->
                result.success(ARYPush.getSubscribedTopics().toList())

            "getInitialNotification" ->
                result.success(ARYPush.consumeInitialNotification()?.toMap())

            "trackEvent" -> {
                ARYPush.trackEvent(
                    call.argument<String>("name").orEmpty(),
                    call.stringMap("properties")
                )
                result.success(null)
            }

            "flush" -> {
                ARYPush.flush()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun MethodCall.stringMap(key: String): Map<String, String> =
        argument<Map<*, *>>(key)
            ?.entries
            ?.mapNotNull { entry ->
                val entryKey = entry.key ?: return@mapNotNull null
                val entryValue = entry.value ?: return@mapNotNull null
                entryKey.toString() to entryValue.toString()
            }
            ?.toMap()
            .orEmpty()

    private fun MethodCall.stringList(key: String): List<String> =
        argument<List<*>>(key)?.mapNotNull { it?.toString() }.orEmpty()

    /**
     * Maps the Kotlin enum onto the value the Dart enum parses.
     *
     * Kotlin uses SCREAMING_SNAKE_CASE and Dart uses lowerCamelCase, so the mapping is explicit
     * rather than derived: a rename on either side should break here, loudly, instead of
     * silently degrading every status to "not determined".
     */
    private fun PushPermissionStatus.toWire(): String = when (this) {
        PushPermissionStatus.NOT_DETERMINED -> "notDetermined"
        PushPermissionStatus.GRANTED -> "granted"
        PushPermissionStatus.DENIED -> "denied"
        PushPermissionStatus.PROVISIONAL -> "provisional"
        PushPermissionStatus.EPHEMERAL -> "ephemeral"
        PushPermissionStatus.RESTRICTED -> "restricted"
    }

    private companion object {
        const val METHOD_CHANNEL = "ary_push/methods"
        const val EVENT_CHANNEL = "ary_push/events"
        const val MAX_PENDING_EVENTS = 50
    }
}

File: flutter/ios/Classes/ARYPushPlugin.swift
// Vendored into this pod by the podspec prepare_command in a Flutter build, and a separate
// module when the SDK is consumed directly. canImport covers both without two source files.
#if canImport(ARYPush)
import ARYPush
#endif
import Flutter
import UIKit

/// The iOS half of the Flutter bridge.
///
/// Deliberately thin. Every decision about APNs, presentation, tokens, storage and deduplication
/// is made by the native SDK; this class translates method calls and forwards events. There is no
/// second notification engine here, because two engines would drift and the Dart one cannot run
/// while the app is terminated, which is exactly when a notification tap has to be captured.
///
/// Engine lifecycle is the subtle part. A hot restart or a recreated `FlutterEngine` tears down
/// the Dart side while the native SDK keeps running, so listeners are attached when Dart starts
/// listening and detached when it stops. Without that, every hot restart would leave another
/// listener behind and the app would show duplicate events that vanish on a cold start.
public class ARYPushPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private static let methodChannelName = "ary_push/methods"
    private static let eventChannelName = "ary_push/events"

    /// Bounded, because a device that receives many messages before the engine attaches must not
    /// accumulate them forever. Notification opens are not buffered here: the native SDK
    /// persists those itself, which is what makes them survive a terminated launch rather than
    /// merely a slow one.
    private static let maxPendingEvents = 50

    private var eventSink: FlutterEventSink?
    private var pendingEvents: [[String: Any]] = []

    private var receivedListenerId: UUID?
    private var openedListenerId: UUID?
    private var tokenListenerId: UUID?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ARYPushPlugin()

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - FlutterStreamHandler

    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        eventSink = events
        attachListeners()

        // Anything that happened while Dart was not listening is delivered now, in order.
        let queued = pendingEvents
        pendingEvents.removeAll()
        queued.forEach { events($0) }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        detachListeners()
        eventSink = nil
        return nil
    }

    private func attachListeners() {
        detachListeners()
        receivedListenerId = ARYPush.addNotificationReceivedListener { [weak self] in
            self?.emit(type: "received", payload: $0.toDictionary())
        }
        openedListenerId = ARYPush.addNotificationOpenedListener { [weak self] in
            self?.emit(type: "opened", payload: $0.toDictionary())
        }
        tokenListenerId = ARYPush.addTokenRefreshListener { [weak self] token in
            self?.emit(type: "token", payload: ["token": token])
        }
    }

    private func detachListeners() {
        if let receivedListenerId {
            ARYPush.removeNotificationReceivedListener(receivedListenerId)
        }
        if let openedListenerId {
            ARYPush.removeNotificationOpenedListener(openedListenerId)
        }
        if let tokenListenerId {
            ARYPush.removeTokenRefreshListener(tokenListenerId)
        }
        receivedListenerId = nil
        openedListenerId = nil
        tokenListenerId = nil
    }

    private func emit(type: String, payload: [String: Any]) {
        let event: [String: Any] = ["type": type, "payload": payload]
        guard let eventSink else {
            if pendingEvents.count >= Self.maxPendingEvents { pendingEvents.removeFirst() }
            pendingEvents.append(event)
            return
        }
        // Channel messages must be sent from the platform thread.
        if Thread.isMainThread {
            eventSink(event)
        } else {
            DispatchQueue.main.async { eventSink(event) }
        }
    }
}

// MARK: - Method calls

extension ARYPushPlugin {

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "initialize":
            ARYPush.initialize(FlutterConfigMapper.from(call.arguments))
            result(nil)

        case "isInitialized":
            result(ARYPush.isInitialized)

        case "requestPermission":
            // Answered asynchronously: the system prompt is shown and the result reported once
            // the user has decided.
            ARYPush.requestPermission { status in result(status.rawValue) }

        case "getPermissionStatus":
            ARYPush.getPermissionStatus { status in result(status.rawValue) }

        case "openNotificationSettings":
            Task { @MainActor in
                ARYPush.openNotificationSettings()
                result(nil)
            }

        case "getInstallationId":
            result(ARYPush.getInstallationId())

        case "getPushToken":
            result(ARYPush.getPushToken())

        case "getPushProvider":
            result(ARYPush.getPushProvider().wireValue)

        case "login":
            guard let userId = arguments["userId"] as? String, !userId.isEmpty else {
                result(
                    FlutterError(
                        code: "invalid_argument",
                        message: "userId must not be blank",
                        details: nil
                    )
                )
                return
            }
            ARYPush.login(userId)
            result(nil)

        case "logout":
            ARYPush.logout()
            result(nil)

        case "getUserId":
            result(ARYPush.getUserId())

        case "addTags":
            ARYPush.addTags(stringMap(arguments["tags"]))
            result(nil)

        case "removeTags":
            ARYPush.removeTags(Set(stringList(arguments["keys"])))
            result(nil)

        case "removeAllTags":
            ARYPush.removeAllTags()
            result(nil)

        case "getTags":
            result(ARYPush.getTags())

        case "subscribeToTopic":
            result(ARYPush.subscribeToTopic(arguments["topic"] as? String ?? ""))

        case "unsubscribeFromTopic":
            result(ARYPush.unsubscribeFromTopic(arguments["topic"] as? String ?? ""))

        case "getSegments":
            // Answered asynchronously: membership is read from the backend.
            ARYPush.getSegments { segments in
                result(segments.map { $0.toDictionary() })
            }

        case "getSubscribedTopics":
            result(Array(ARYPush.getSubscribedTopics()).sorted())

        case "getInitialNotification":
            result(ARYPush.consumeInitialNotification()?.toDictionary())

        case "trackEvent":
            ARYPush.trackEvent(
                arguments["name"] as? String ?? "",
                properties: stringMap(arguments["properties"])
            )
            result(nil)

        case "flush":
            ARYPush.flush()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func stringMap(_ value: Any?) -> [String: String] {
        guard let dictionary = value as? [String: Any] else { return [:] }
        return dictionary.reduce(into: [String: String]()) { result, entry in
            result[entry.key] = String(describing: entry.value)
        }
    }

    private func stringList(_ value: Any?) -> [String] {
        (value as? [Any])?.map { String(describing: $0) } ?? []
    }
}

/// Translates the Dart configuration map into a native ``ARYPushConfig``.
///
/// Every field is optional and every unrecognised value falls back to the native default. A
/// configuration mistake in Dart should degrade to "the SDK ran with defaults", never to a failed
/// initialization: an application that cannot start because a log level was misspelled is a worse
/// outcome than one that logs at the wrong level.
enum FlutterConfigMapper {

    static func from(_ arguments: Any?) -> ARYPushConfig {
        guard let map = arguments as? [String: Any] else { return ARYPushConfig() }

        var backend: PushBackendConfig?
        if let backendMap = map["backend"] as? [String: Any],
           let baseURL = backendMap["baseUrl"] as? String,
           !baseURL.isEmpty {
            backend = PushBackendConfig(
                baseURL: baseURL,
                applicationId: backendMap["applicationId"] as? String,
                apiVersion: backendMap["apiVersion"] as? String ?? "v1"
            )
        }

        return ARYPushConfig(
            enableLogging: map["enableLogging"] as? Bool ?? false,
            logLevel: logLevel(map["logLevel"] as? String) ?? .info,
            autoRequestPermission: map["autoRequestPermission"] as? Bool ?? false,
            foregroundDisplay: foregroundPolicy(map["foregroundDisplay"] as? String) ?? .show,
            backend: backend,
            collectDeviceInfo: map["collectDeviceInfo"] as? Bool ?? true
        )
    }

    private static func logLevel(_ value: String?) -> PushLogLevel? {
        switch value?.lowercased() {
        case "verbose": return .verbose
        case "debug": return .debug
        case "info": return .info
        case "warning": return .warning
        case "error": return .error
        case "none": return PushLogLevel.none
        default: return nil
        }
    }

    private static func foregroundPolicy(_ value: String?) -> ForegroundDisplayPolicy? {
        switch value?.lowercased() {
        case "show": return .show
        case "eventonly": return .eventOnly
        case "suppress": return .suppress
        default: return nil
        }
    }
}
